### ISSUE TITLE
Edward support for recent extensions of VAE 

### ISSUE THREAD

**Issue by rfarouni** [blei-lab/edward#447]
@dustinvtran  Hi Dustin,

How hard do you think it is to implement the following models in Edwards using the Tensorflow backend?

- Hierarchical variational models
- [Variational Recurrent Neural Network](https://arxiv.org/abs/1506.02216)
- [Discrete Variational Autoencoders](https://arxiv.org/abs/1609.02200)
- [Deep Markov models](https://arxiv.org/abs/1609.09869)

Is there a class of models that Edward won't be able to handle in the near future?

Thanks

**Comment by dustinvtran** [comment]
we designed the library specifically so those examples are as easy to do as possible.

i think in general, there will never be a class of models that Edward won't be able to handle. however, as with all software design, there are always models/inference that are easier or harder to work with. for example, i think bayesian nonparametrics and undirected models are possible in edward—arguably they are best done in edward as there's no other library that captures them—but they will always be not as easy to do as simpler, directed graphical models.

**Comment by rfarouni** [comment]
That's great! So for implementing HVM on tensorflow for example, do you recommend I modify the example [here](https://github.com/blei-lab/edward/blob/original/examples/gaussian_hvm.py).

**Comment by dustinvtran** [comment]
the easiest approach to get HVMs is to not write a new algorithm altogether. rather you can just use `KLqp` on the joint space, where you minimize

KL ( q(z, lambda; theta) || r(lambda; phi)p(z | x) ).

here, q(z, lambda; theta) is the hierarchical variational model, and r(lambda; phi) is the auxiliary model. this doesn't require a new algorithm and in fact is exactly equivalent to the HVM algorithm (up to choices of Rao-Blackwellization and mixes of reparameterization vs score function gradient). for example,
```python
# probability model
z = RandomVariable(..)
x = RandomVariable(..)

# (hierarchical) variational model
lambda = RandomVariable(..)
z = RandomVariable(..)

# auxiliary model
rlambda = RandomVariable(..)

inference = ed.KLqp({z: qz, rlambda: qlambda}, data={x: x_data})
```
more details are in the paragraph below Equation 5 in the [paper](http://dustintran.com/papers/RanganathTranBlei2016.pdf).

**Comment by rfarouni** [comment]
That helps a lot. Thank you for your help!

**Comment by yunhuiguo** [comment]
@dustinvtran Hi Dustin,

I am still confused about how to use HVM in Edward. Could please provide a complete tutorial or an example on this? Thanks!

**Comment by XinDongol** [comment]
@dustinvtran I am looking for example of HVM (in ICML) too.
